io.c

```diff
@@ -10922,62 +10918,82 @@ pipe_pair_close(VALUE rw)
 
 /*
  *  call-seq:
- *     IO.pipe                             ->  [read_io, write_io]
- *     IO.pipe(ext_enc)                    ->  [read_io, write_io]
- *     IO.pipe("ext_enc:int_enc" [, opt])  ->  [read_io, write_io]
- *     IO.pipe(ext_enc, int_enc [, opt])   ->  [read_io, write_io]
+ *    IO.pipe(enc_string = 'UTF-8', **opts) -> [read_io, write_io]
+ *    IO.pipe(enc_string = 'UTF-8', **opts) {|read_io, write_io] ...} -> object
+ *    IO.pipe(int_enc = Encoding::UTF_8, ext_enc = nil, **opts) -> [read_io, write_io]
+ *    IO.pipe(int_enc = Encoding::UTF_8, ext_enc = nil, **opts) {|read_io, write_io] ...} -> object
  *
- *     IO.pipe(...) {|read_io, write_io| ... }
+ *  Creates a pair of pipe endpoints, +read_io+ and +write_io+,
+ *  connected to each other.
  *
- *  Creates a pair of pipe endpoints (connected to each other) and
- *  returns them as a two-element array of IO objects:
- *  <code>[</code> <i>read_io</i>, <i>write_io</i> <code>]</code>.
+ *  If argument +enc_string+ is given, it must be a string containing one of:
  *
- *  If a block is given, the block is called and
- *  returns the value of the block.
- *  <i>read_io</i> and <i>write_io</i> are sent to the block as arguments.
- *  If read_io and write_io are not closed when the block exits, they are closed.
- *  i.e. closing read_io and/or write_io doesn't cause an error.
+ *  - The name of the encoding to be used as the internal encoding.
+ *  - The colon-separated names of two encodings to be used as the internal
+ *    and external encodings.
  *
- *  Not available on all platforms.
+ *  An array of the encoding names is returned by method Encoding.name_list.
+ *
+ *  If argument +int_enc+ is given, it must be an encoding object
+ *  that specifies the internal encoding to be used;
+ *  if argument +ext_enc+ is also given, it must be an encoding object
+ *  that specifies the external encoding to be used.
+ *
+ *  An array of encoding classes is returned by Encoding.list.
+ *
+ *  The string read from +read_io+ is tagged with the internal encoding;
+ *  if an external encoding is also specified, the string is converted
+ *  to, and tagged with, that encoding.
```

Comment:
I think it is confused, it should be other way round (external encoding is what incoming data is tagged with, internal encoding is what it converted into).

Also, maybe we should provide an example demonstrating the encodings? Here is my update of the original example with using Ukrainian encoding KOI8-U and Ukrainian text (it means "hi"):

```ruby
rd, wr = IO.pipe('KOI8-U:UTF-8')

if fork
  wr.close
  str = rd.read
  puts "Parent got: <#{str}/#{str.encoding}>"
  rd.close
  Process.wait
else
  rd.close
  puts 'Sending message to parent'
  wr.write "привіт".encode('KOI8-U')
  wr.close
end
```
With various changes of the encoding parameter, the effect of it on the result can be clarified.

Reply:
As I said earlier, I want to do more with the encodings here; that can include examples with their effects.  Can we leave that for the next PR?

Reply:
Of course, we can do it later, if in the current one we'll just reverse this part of the changes.